e2e/e2e_node_slice_test.go

```diff
@@ -0,0 +1,353 @@
+
+			By("creating a NetworkAttachmentDefinition for whereabouts")
+			_, err = clientInfo.AddNetAttachDef(netAttachDef)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("checking node slices have been allocated and nodes are assigned")
+			Expect(util.ValidateNodeSlicePoolSlicesCreatedAndNodesAssigned(testNetworkName, testNamespace, subnets, clientInfo)).To(Succeed())
+		})
+
+		AfterEach(func() {
+			Expect(clientInfo.DelNetAttachDef(netAttachDef)).To(Succeed())
+			time.Sleep(1 * time.Second)
+			Expect(clientInfo.NodeSliceDeleted(testNetworkName, testNamespace)).To(Succeed())
+		})
+
+		Context("Single pod tests node slice", func() {
+			BeforeEach(func() {
+				const singlePodName = "whereabouts-basic-test"
+				var err error
+
+				By("creating a pod with whereabouts net-attach-def")
+				pod, err = clientInfo.ProvisionPod(
+					singlePodName,
+					testNamespace,
+					util.PodTierLabel(singlePodName),
+					entities.PodNetworkSelectionElements(testNetworkName),
+				)
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			AfterEach(func() {
+				By("deleting pod with whereabouts net-attach-def")
+				Expect(clientInfo.DeletePod(pod)).To(Succeed())
+			})
+
+			It("allocates a single pod within the correct IP range", func() {
+				By("checking pod IP is within whereabouts IPAM range")
+				secondaryIfaceIPs, err := retrievers.SecondaryIfaceIPValue(pod)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(secondaryIfaceIPs).NotTo(BeEmpty())
+				Expect(util.InNodeRange(clientInfo, pod.Spec.NodeName, testNetworkName, testNamespace, secondaryIfaceIPs[0])).To(Succeed())
+			})
+		})
+
+		Context("Replicaset tests node slice", func() {
+			const (
+				emptyReplicaSet = 0
+				rsSteadyTimeout = 1200 * time.Second
+			)
+
+			var k8sIPAM *wbstorage.KubernetesIPAM
+
+			BeforeEach(func() {
+				By("creating a replicaset with whereabouts net-attach-def")
+				var err error
+
+				const ipPoolNamespace = "kube-system"
+				k8sIPAM, err = wbstorage.NewKubernetesIPAMWithNamespace("", types.IPAMConfig{
+					Kubernetes: types.KubernetesConfig{
+						KubeConfigPath: testConfig.KubeconfigPath,
+					},
+					NodeSliceSize: sliceSize,
+					NetworkName:   testNetworkName,
+					Namespace:     testNamespace,
+				}, ipPoolNamespace)
+				Expect(err).NotTo(HaveOccurred())
+
+				replicaSet, err = clientInfo.ProvisionReplicaSet(
+					rsName,
+					testNamespace,
+					emptyReplicaSet,
+					util.PodTierLabel(rsName),
+					entities.PodNetworkSelectionElements(testNetworkName),
+				)
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			AfterEach(func() {
+				By("removing replicas and expecting 0 IP pool allocations")
+				Expect(
+					util.CheckZeroIPPoolAllocationsAndReplicas(
+						context.TODO(), clientInfo, k8sIPAM, rsName, testNamespace, ipv4TestRange, testNetworkName)).To(Succeed())
+
+				By("deleting replicaset with whereabouts net-attach-def")
+				Expect(clientInfo.DeleteReplicaSet(replicaSet)).To(Succeed())
+			})
+
```

Comment:
Can these tests be run from multi-threaded in parallel, to verify simultaneous requests from several client?

Reply:
We could do something like that but what exactly are we trying to test with that? Whereabouts works on pod create so these tests do launch many concurrent pods.

Reply:
If I understand correctly the pod creation here is serial, so the IP requests toward whereabouts may be not fully concurrent, and so don't test simultaneous requests (at least if the test is blocking on pod creation)

Reply:
So it sets the replicaset to `testConfig.MaxReplicas(allPods.Items)` that will result in many pods launching in parallel. Depending on how many nodes are in the test cluster this should also lead to multiple pods per node so this would get exercised with multiple pods to a lease/node pool.